ZScript RPG system review

MorthimerMcMare opened the pull request:
Итак, RPG-система готова. Осталось дело за малым: отладить...


__Review__:
- [x] Требуется слишком много опыта для перехода на следующий уровень
- [x] **В опыт не учитывается ломание предметов/тушение огня/открытие заклинивших дверей/открывание сундуков/взрывание бочек**
- [x] В окошке отображается предыдущий уровень, а не тот, на который игрок только что перешел
- [ ] Слишком читерский урон при прокачке силы, хотя бы, до 8
- [x] Собаки не берут урон на себя
- [x] Топор не блокирует часть урона, так же, нет звука блокировки

MorthimerMcMare committed:
RPG-система на ZScript. 

Работает всё (насколько я протестировал; исключая связанное с бронёй) -- от повышения уровня до магазинов. 

Присутствуют следующие ACS-вызовы: "int GetFeature( Actor activator, int type )" и "void SetFeature( Actor activator, int type, int value )". 

Также повышать/понижать количественную характеристику атрибутов можно через "Actor DRRPShopPlayer<feature_name>: Inventory", просто выставляя нужное количество этого предмета (для понижения параметра нужно установить бит 0x100 (256), младшие восемь -- без изменений, но сколько вычитать, не прибавлять).

MorthimerMcMare committed:
Дополнение StatInfo-меню параметрами ZScript RPG-сисиемы.

Теперь помимо уже сделанного показывается уровень, опыт и четыре основных атрибута.

PROPHESSOR commented:
![screenshot](https://media.discordapp.net/attachments/505122055935623199/559856752091463680/unknown.png)
Уже дошел сюда, а повышения уровня всё не было... Обычно, оно происходит перед дверью 1234

PROPHESSOR commented:
![screenshot](https://media.discordapp.net/attachments/505122055935623199/559857008786931712/unknown.png)
У меня уже второй уровень

MorthimerMcMare commented:
Понимаешь, у тебя ящики и огонь тоже учитывались, а я их пока отключил.

PROPHESSOR commented:
Не надо их отключать

PROPHESSOR commented:
![image](https://user-images.githubusercontent.com/19251320/54956907-eab62500-4f59-11e9-8e1d-e7516873118e.png)
Ничего себе! Нужно 54 монстра убить, чтобы получить 3 уровень?!

MorthimerMcMare commented:
Ну да. Пока так.

Всё-таки вскоре несколько опыта сразу даваться за убийство будет.

PROPHESSOR commented:
На третьем уровне все монстры первого уровня умирают с одного-двух выстрелов пистолета...

PROPHESSOR commented:
Собаки не берут урон на себя

PROPHESSOR commented:
Защита не учитывается при получении урона
![image](https://user-images.githubusercontent.com/19251320/54957931-d293d500-4f5c-11e9-9098-f0feb9489529.png)

PROPHESSOR commented:
Блокирование урона топором не работает (ни на урон не влияет, ни звук не проигрывается)

MorthimerMcMare committed:
Исправление по мотивам "Request changes" от PROPHESSOR'а.

Вкратце: https://trello.com/c/Oqg9EMtp/ .

PROPHESSOR commented:
Обобщу:
- Требуется слишком много опыта для перехода на следующий уровень
- В опыт не учитывается ломание предметов/тушение огня/открытие заклинивших дверей/открывание сундуков/взрывание бочек
- В окошке отображается предыдущий уровень, а не тот, на который игрок только что перешел
- Слишком читерский урон при прокачке силы, хотя бы, до 8
- Собаки не берут урон на себя
- Топор не блокирует часть урона, так же, нет звука блокировки

PROPHESSOR reviewed (changes requested):
Ах да, я забыл запустить Review)

MorthimerMcMare commented:
Исправил всё, что не относится к балансу (фактические ошибки).

MorthimerMcMare commented:
Остальное нужно подбирать долгим путём. Баланс относится к RPG-системе, но не к фундаменту оной, и считать его ошибки фатальными (запрещающими делать "Merge pull request") как минимум странно, как максимум -- глупо.

PROPHESSOR commented:
```Исправил всё, что не относится к балансу (фактические ошибки).```
Увы, нет. Я отметил выполненные пункты (см. описание PR'а)

```Остальное нужно подбирать долгим путём. Баланс относится к RPG-системе, но не к фундаменту оной, и считать его ошибки фатальными (запрещающими делать "Merge pull request") как минимум странно, как максимум -- глупо.```
Не соглашусь. В ветке master должна быть более-менее стабильная версия.
Т.е. нужно прикинуть баланс, чтобы он был в разумных пределах, а не оставлять надежду под конец игры получить 10 уровень.

MorthimerMcMare commented:
`Не соглашусь. В ветке master должна быть более-менее стабильная версия. Т.е. нужно прикинуть баланс, чтобы он был в разумных пределах, а не оставлять надежду под конец игры получить 10 уровень.`
Ну... Опять же, не соглашусь. В твоей версии баланса также не было, однако этот ACS-код был спокойно отправлен в `master` (все значения урона, защиты и так далее брал с твоего кода; изменил только количество опыта, требуемое для перехода на следующий уровень). Возможно, также у тебя баланс казался более верным из-за [бага при смене карты](https://trello.com/c/mG7WE3eO/), но это засчитывать за плюс будет верхом "Это не баг, это фича" (напомню, что фактический уровень и значения атрибутов игрока у тебя посмотреть было нельзя, в StatInfo они не выводились).

`Увы, нет. Я отметил выполненные пункты (см. описание PR'а)`
Всё, что не относится к балансу -- выполнил. По остальному поставлен вопрос насчёт твоего ACS-баланса (см. первый пункт).

PROPHESSOR commented:
Выделил полужирным пункт, без которого не пущу =D

MorthimerMcMare commented:
Уговорил. Сделаю небольшой баланс. Также добавлю разное количество опыта за каждого из монстров (дай примерную таблицу).

PROPHESSOR commented:
> дай примерную таблицу

Я знаю не больше твоего... В оригинале такого не было, соответственно, необходимо придумать самим

MorthimerMcMare committed:
Дополнение RPG-системы (точность, опыт за ящики/огонь/...).

Теперь точность всего оружия, за исключением огнетушителя и вещей ближнего боя, зависит от параметра "Accuracy".

Опыт даётся игроку за все разрушения/убийства объектов, пока без разбора.

PROPHESSOR commented:
Принято
